<center>
<img src="images/cover.svg" style="width: 100%;" />
</center>

<center>
<img src="images/how.svg" style="width: 100%;" />
</center>

<center>
<img src="images/pillars.svg" style="width: 100%;" />
</center>

 Costruiamo una IA generativa a partire da 20'000 frasi prese a caso dal Web sul tema cucina.

In [1]:
import brightwrite

esempi = brightwrite.carica_esempi("data/cucina.txt")

brightwrite.stampa_esempi(esempi, 10)

  0%|          | 0/20000 [00:00<?, ?it/s]

1	Questa fase di riposo a freddo è fondamentale , perchè permette alle d...
2	Prima di tirare fuori la pallotta ho raccolto e lavato un rametto di r...
3	Ciao Db sono reduce da una piacevole serata con amici e per l occasion...
4	Purtroppo anche qui trovo indicati mille modi per rovinare questa piet...
5	Riprendete il disco di brisé più grande e stendetelo con l uso di un m...
6	Stiamo parlando di una pianta dalla quale si ricava una farina ad alto...
7	Lasciate cuocere il sugo leggermente scoperto , così evaporerà parte d...
8	In poche righe è praticamente impossibile condensare tutte le peculiar...
9	Di più posso solo dirvi che è un grano di una bontà incredibile , ho d...
10	Nn vi aspettate un pane alveolato con occhiatura ecc . ecc . , nn cred...


Costruiamo il **vocabolario** del nostro linguaggio:
* l'elenco di tutte le parole, e simboli, che compaiono almeno una volta negli esempi

In [ ]:
vocabolario = brightwrite.costruisci_vocabolario(esempi)

Dai testi di esempio possiamo costruire un **modello del linguaggio**.

Cominciamo costruendo un **vocabolario** con le **frequenze** di uso delle parole.

In [2]:
vocabolario, frequenze = brightwrite.conta_frequenze(esempi)
print('dimensione vocabolario:',len(vocabolario))
print()
for idx in range(40):
    print(f'{frequenze[idx]:5} {vocabolario[idx]:15}', end='\t')


dimensione vocabolario: 39696

50576 ,              	35348 .              	33190 di             	29431 e              	19196 NUM            	18288 la             	16766 il             	14313 a              	13549 farina         	13012 per            	12216 un             	12134 con            	11458 in             	 9153 una            	 8440 che            	 7791 '              	 6806 l              	 6526 le             	 6090 cuocere        	 5502 è              	 5263 i              	 5077 non            	 4292 da             	 4231 :              	 4229 si             	 4102 minuti         	 3666 acqua          	 3478 al             	 3461 del            	 3388 della          	 3214 o              	 3110 sale           	 3105 zucchero       	 2945 lo             	 2701 impasto        	 2685 olio           	 2525 lievito        	 2471 circa          	 2443 anche          	 2407 più            	

<center><img src="images/dart.png" style="display:inline-block; float:left;margin-top:10%; width:20%; visibility: hidden; " /><img src="images/scelta_pesata.png" style="margin-top:5%; width:45%"/> </center>


<center><img src="images/dart.png" style="display:inline-block; float:left;margin-top:10%; width:20%" />
    <img src="images/scelta_pesata.png" style="margin-top:5%; width:45%"/> </center>

Scegliamo parole **a caso** con una **probabilità proporzionale alla frequenza d'uso negli esempi**

In [4]:
import random
print(' '.join(random.choices(vocabolario, k=15, weights=frequenze)))

di conosce ma la e addensato , il alta caldo e frigge che ' ,


<center>
    <img src="images/context0.png" />
</center>

<center>
    <img src="images/context1.png" />
</center>

Calcolo le frequenze per tutti i contesti osservati negli esempi:
<center>
    <img src="images/modello.svg" style="width:80%;">
</center>

Generiamo il testo usando frequenze e contesti:
<center>
    <img src="images/gen_chain.svg" style="width:80%;">
</center>

Calcolo le frequenze per tutti i contesti osservati negli esempi:

In [5]:
modello = brightwrite.conta_frequenze_nei_contesti(esempi)

  0%|          | 0/20000 [00:00<?, ?it/s]

Numero contesti: 85282
Numero totale di parametri: 655441


In [6]:
brightwrite.mostra_frequenze_per_contesto(modello, 'mescolare il')

   12 tutto                              5 burro                              1 fior                               1 latte                              1 cocco                              1 caramello                          1 grano                              1 purè                               1 resto                              1 tuorlo                           ... (+5)


In [7]:
brightwrite.mostra_frequenze_per_contesto(modello, 'il cioccolato')

   18 fuso                              13 fondente                          10 e                                  9 .                                  9 con                                8 a                                  8 sciolto                            7 ,                                  5 in                                 5 bianco                           ... (+26)


In [12]:
print(brightwrite.genera_testo(modello, contesto='impastare il'))

impastare il nostro impasto e lievitazione e conferire forza è una cosa un composto omogeneo .


In [13]:
print(brightwrite.genera_testo(modello, contesto='mescolare il', dettagli=True))

Contesto: "mescolare il"	30 opzioni	-> "burro"
Contesto: "mescolare il burro"	5 opzioni	-> "freddo"
Contesto: "il burro freddo"	61 opzioni	-> "di"
Contesto: "burro freddo di"	8 opzioni	-> "frigorifero"
Contesto: "freddo di frigorifero"	5 opzioni	-> "e"
Contesto: "di frigorifero e"	16 opzioni	-> "non"
Contesto: "frigorifero e non"	307 opzioni	-> "risulti"
Contesto: "e non risulti"	20 opzioni	-> "omogenea"
Contesto: "non risulti omogenea"	99 opzioni	-> "."

mescolare il burro freddo di frigorifero e non risulti omogenea .


<center>
    <img src="images/wizard1.svg" style="width:80%;">
</center>

<center>
    <img src="images/wizard2.svg" style="width:80%;">
</center>

<table>
<thead>
  <tr>
    <th></th>
    <th>BrightWrite</th>
    <th>GPT-3</th>
    <th>GPT-4</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Dimensione esempi</td>
    <td>800 mila parole</td>
    <td>300 miliardi</td>
    <td>13000 miliardi</td>
  </tr>
  <tr>
    <td>Dimensione contesto</td>
    <td>3 parole</td>
    <td>2048</td>
    <td>128000</td>
  </tr>
  <tr>
    <td>Numero parametri<br>nel modello</td>
    <td>655441</td>
    <td>175 miliardi</td>
    <td>1800 miliardi</td>
  </tr>
  <tr>
    <td>Costo di produzione</td>
    <td><0.01 \$</td>
    <td>4.6 M\$</td>
    <td>100 M\$</td>
  </tr>
</tbody>
</table>

<center>
    <img src="images/chat1.png" style="width:80%;">
</center>

<center>
    <img src="images/chat2.png" style="width:80%;">
</center>

<center>
    <img src="images/chat3.png" style="width:80%;">
</center>

<center>
    <img src="images/image_gen_0.svg" width="80%"/>
</center>

<center>
    <img src="images/image_gen_1.svg" width="80%"/>
</center>

<center>
    <img src="images/image_gen_2.svg" width="80%"/>
</center>

<center>
    <img src="images/image_gen_3.svg" width="80%"/>
</center>

<center>
    <img src="images/image_gen_5.svg" width="80%"/>
</center>

<img src="images/qrcode.png" style="width:20%;float:left;"/>

Codice open source: 

https://github.com/aesuli/brightwrite

Immagini create con DALL-E 3.

Esempi ChatGPT creati con la versione 3.5.